File: Tutorial/NDOTestApp/NDOTravelExpenses-Step 6 VB/BusinessClasses/Picture.vb
Imports NDO
Imports System
Imports System.Drawing
Imports System.Drawing.Imaging
Imports System.IO


<NDOPersistent()> _
Public Class Picture
    Inherits PictureHeader

    ' Methods
    Public Sub New()
        _rawbytes = Nothing
        _image = Nothing
    End Sub


    ' Properties
    Public Property Image() As Image
        Get
            If (_rawbytes Is Nothing) Then
                Return Nothing
            End If
            If (_image Is Nothing) Then
                Dim stream1 As New MemoryStream(_rawbytes)
                _image = System.Drawing.Image.FromStream(stream1)
            End If
            Return _image
        End Get
        Set(ByVal value As Image)
            _image = value
            If (_image Is Nothing) Then
                _rawbytes = Nothing
            Else
                ' align horizontal size to 4 bytes 
                Dim horsize As Integer = _image.Size.Width * 3
                Dim remainder As Integer = horsize Mod 4
                If (remainder > 0) Then
                    remainder = 1
                End If
                horsize = horsize \ 4
                horsize = (horsize + remainder) * 4
                Dim vertsize As Integer = _image.Size.Height
                _rawbytes = New Byte(horsize * vertsize + 56) {}
                Dim stream1 As New MemoryStream(_rawbytes)
                _image.Save(stream1, ImageFormat.Bmp)
            End If
        End Set
    End Property


    ' Fields
    <NDOTransient()> _
    Private _image As Image
    Private _rawbytes As Byte()
End Class

File: Tutorial/NDOTestApp/NDOTravelExpenses-Step 6/BusinessClasses/Picture.cs
using System;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using NDO;

namespace BusinessClasses
{
    [NDOPersistent]
    public partial class Picture : PictureHeader
    {
        byte[] rawbytes = null;
        [NDOTransient]
        // can not be stored directly
        Image image = null;
        public Image Image
        {
            get
            {
                if (rawbytes == null)
                    return null;
                if (image == null)
                {
                    System.IO.MemoryStream ms = new MemoryStream(rawbytes);
                    image = Image.FromStream(ms);
                }
                return image;
            }
            set
            {
                image = value;
                if (image == null)
                {
                    rawbytes = null;
                    return;
                }
                // horizontal align to 4 bytes 
                int horsize = image.Size.Width * 3;
                int rest = horsize % 4;
                if (rest > 0) rest = 1;
                horsize = horsize / 4;
                horsize = (horsize + rest) * 4;
                int vertsize = image.Size.Height;
                rawbytes = new byte[horsize * vertsize + 56];
                System.IO.MemoryStream ms = new MemoryStream(rawbytes);
                image.Save(ms, ImageFormat.Bmp);
            }
        }
        public Picture()
        {
        }
    }
}
